processing/src/main/java/io/druid/query/groupby/epinephelinae/ParallelCombiner.java

```diff
@@ -0,0 +1,536 @@
+    final ResourceHolder<ByteBuffer> combineBufferHolder = combineBufferSupplier.get();
+    final ByteBuffer combineBuffer = combineBufferHolder.get();
+    final int minimumRequiredBufferCapacity = StreamingMergeSortedGrouper.requiredBufferCapacity(
+        combineKeySerdeFactory.factorizeWithDictionary(mergedDictionary),
+        combiningFactories
+    );
+    // We want to maximize the parallelism while the size of buffer slice is greater than the minimum buffer size
+    // required by StreamingMergeSortedGrouper. Here, we find the leafCombineDegree of the cominbing tree and the
+    // required number of buffers maximizing the parallelism.
+    final Pair<Integer, Integer> degreeAndNumBuffers = findLeafCombineDegreeAndNumBuffers(
+        combineBuffer,
+        minimumRequiredBufferCapacity,
+        concurrencyHint,
+        sortedIterators.size()
+    );
+
+    final int leafCombineDegree = degreeAndNumBuffers.lhs;
+    final int numBuffers = degreeAndNumBuffers.rhs;
+    final int sliceSize = combineBuffer.capacity() / numBuffers;
+
+    final Supplier<ByteBuffer> bufferSupplier = createCombineBufferSupplier(combineBuffer, numBuffers, sliceSize);
+
+    final Pair<List<CloseableIterator<Entry<KeyType>>>, List<Future>> combineIteratorAndFutures = buildCombineTree(
+        sortedIterators,
+        bufferSupplier,
+        combiningFactories,
+        leafCombineDegree,
+        mergedDictionary
+    );
+
+    final CloseableIterator<Entry<KeyType>> combineIterator = Iterables.getOnlyElement(combineIteratorAndFutures.lhs);
+    final List<Future> combineFutures = combineIteratorAndFutures.rhs;
+
+    final Closer closer = Closer.create();
+    closer.register(combineBufferHolder);
+    closer.register(() -> checkCombineFutures(combineFutures));
+
+    return CloseableIterators.wrap(combineIterator, closer);
+  }
+
+  private static void checkCombineFutures(List<Future> combineFutures)
+  {
+    for (Future future : combineFutures) {
+      try {
+        if (!future.isDone()) {
+          // Cancel futures if close() for the iterator is called early due to some reason (e.g., test failure)
+          future.cancel(true);
+        } else {
+          future.get();
+        }
+      }
+      catch (InterruptedException | CancellationException e) {
+        throw new QueryInterruptedException(e);
+      }
+      catch (ExecutionException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  private static Supplier<ByteBuffer> createCombineBufferSupplier(
+      ByteBuffer combineBuffer,
+      int numBuffers,
+      int sliceSize
+  )
+  {
+    return new Supplier<ByteBuffer>()
+    {
+      private int i = 0;
+
+      @Override
+      public ByteBuffer get()
+      {
+        if (i < numBuffers) {
+          return Groupers.getSlice(combineBuffer, sliceSize, i++);
+        } else {
+          throw new ISE("Requested number of buffer slices exceeds the planned one");
```

Comment:
Let's add the number of requested buffer slices and the planned number for better debugging

Reply:
Added.